taichi/analysis/gather_uniquely_accessed_pointers.cpp

```diff
@@ -172,7 +226,60 @@ class UniquelyAccessedSNodeSearcher : public BasicStmtVisitor {
     }
   }
 
-  static std::unordered_map<const SNode *, GlobalPtrStmt *> run(IRNode *root) {
+  void visit(ExternalPtrStmt *stmt) override {
+    // A memory location of an ExternalPtrStmt depends on the indices
+    // If the accessed indices are loop unique,
+    // the accessed memory location is loop unique
+    for (auto base_ptr : stmt->base_ptrs.data) {
+      ArgLoadStmt *arg_load_stmt = base_ptr->as<ArgLoadStmt>();
+      int arg_id = arg_load_stmt->arg_id;
+
+      auto accessed_ptr = accessed_arr_pointer_.find(arg_id);
+
+      bool stmt_loop_unique =
+          loop_unique_stmt_searcher_.is_ptr_indices_loop_unique(stmt);
+
+      if (!stmt_loop_unique) {
+        accessed_arr_pointer_[arg_id] = nullptr;  // not loop-unique
+      } else {
+        if (accessed_ptr == accessed_arr_pointer_.end()) {
+          // First time using arr @ arg_id
+          accessed_arr_pointer_[arg_id] = stmt;
+        } else {
+          /**
+           * We know stmt->base_ptr and the previously recorded pointers
+           * are loop-unique. We need to figure out whether their loop-unique
+           * indicies are the same while ignoring the others.
+           * e.g. a[i, j, 1] and a[i, j, 2] are both uniquely accessed
+           *      a[i, j, 1] and a[j, i, 2] are not uniquely accessed
+           *      a[i, j + 1, 1] and a[i, j, 2] are not uniquely accessed
+           */
+          if (accessed_ptr->second) {
+            ExternalPtrStmt *other_ptr = accessed_ptr->second;
+            TI_ASSERT(stmt->indices.size() == other_ptr->indices.size());
+            for (int axis = 0; axis < stmt->indices.size(); axis++) {
+              Stmt *this_index = stmt->indices[axis];
+              Stmt *other_index = other_ptr->indices[axis];
+              // We only compare unique indices here.
+              // Since both pointers are loop-unique, all the unique indices
+              // need to be the same for both to be uniquely accessed
```

Comment:
I think we should also care about other indices... e.g., what if we access `a[i, j, 1]` and `a[3, 2, 1]`? There can be a data race when the `i = 3, j = 2` thread is accessing `a[i, j, 1]` when another thread is accessing `a[3, 2, 1]`.

Reply:
That won't pass this test since [3,2,1] is not loop unique (just loop invariant). Once we have one of them being not loop unique, we will have the access to this NDarray marked as not unique.

Reply:
And for loop unique stmts, we are checking that the unique indicies of those statements have the same value, so if two already loop unique pointer uses the same index for them, they should be uniquely accessed as well

Reply:
Makes sense, now I see what's happening here. I think we should rename `is_loop_unique` to something like `is_index_partially_loop_unique` or some better names, because "loop-unique" should refer to the accesses like `a[i, j, 1]`, and `i` itself is not loop-unique (`i` combined with `j` is loop-unique).

BTW, this is a little bit conservative in some corner cases: suppose we have accesses `a[i, j, i]` and `a[i, j, j]`. We can demote the atomics but it's hard to see. If we really want to demote them, we can maintain the set of identical loop-unique indices' positions (indices 0 and 1 in this case, which are `i` and `j`). Then if we have a third access `a[i, i, j]`, we can see that we cannot demote the atomics because although having both `a[i, j, j]` and `a[i, i, j]` is good, the set of identical loop-unique index's position is only index 0 (which is `i`), which is not enough for the whole access to be loop-unique.